src/Common/LRUCache.h

```diff
@@ -276,16 +332,24 @@ class LRUCache
         return cell.value;
     }
 
-    void setImpl(const Key & key, const MappedPtr & mapped, [[maybe_unused]] std::lock_guard<std::mutex> & cache_lock)
+    bool setImpl(const Key & key, const MappedPtr & mapped, [[maybe_unused]] std::lock_guard<std::mutex> & cache_lock)
     {
         auto [it, inserted] = cells.emplace(std::piecewise_construct,
             std::forward_as_tuple(key),
             std::forward_as_tuple());
 
         Cell & cell = it->second;
+        auto value_weight = mapped ? weight_function(*mapped) : 0;
 
         if (inserted)
         {
+            if (!removeOverflow(value_weight))
+            {
+                // cannot find enough space to put in the new value
+                cells.erase(it);
+                return false;
+            }
+
             try
             {
                 cell.queue_iterator = queue.insert(queue.end(), key);
```

Comment:
Also this changes previous behaviour in case in try catch below we get an exception. I mean now we could remove some values with `removeOverflow`, then get exception and not insert anything, but we already made place for this insertion. (Even though this is an exception from insertion into list and not in general likely to happen). So better to first insert into queue, remember iterator to currently inserted key, then check if there is enough space, it not -- remove from queue.

Reply:
done

Reply:
It was a bad idea actually. Or now we need to guarantee that for key, which we want to set, `canRelease` must return true, and also one more problem -- `release()` (is it legal to call it if in fact we did not return from `set`?). For cache it removes files from local fs, and BTW does not currently check if path exists before doing that.

I'll think what to do with that, how to do it differently. If you also have ideas -- propose.
Also will need to add unit tests for changes from this pr.

Reply:
Ok no problem here after all because value will be `nullptr`, but then need to move check for nullptr out of `release()` and `canRelease()` and pass a reference inside.

Reply:
btw, I don't find where to add a unit test

Reply:
Similar to `src/Common/tests/gtest_lru_hash_map.cpp`
to run:
```
 ninja unit_tests_dbms
 cd build/src
./unit_tests_dbms --gtest_filter='TransformQueryForExternalDatabase*' --gtest_brief=0 --gtest_throw_on_failure=1 
```

Reply:
Another fatal bug, when `canRelease()` returns true, before calling `release()`, the element may turn into unreleasable again

Reply:
> Another fatal bug, when canRelease() returns true, before calling release(), the element may turn into unreleasable again

As I understood you, what you say should not happen because we hold mutex in `set` and in `get`. It may become unreleasable if someone successfully called `get`, but while we are in `trySet` this will not happen.

Reply:
following is the case
1. we get A from cache, A is releasable
2. then we try to set B and sees A is releasable, 
3. at now something happpens on A outside lru cache, and make A unreleasable
4. then lru cache call release A to make space for B.

Reply:
This is the case I asked you to fix here https://github.com/ClickHouse/ClickHouse/pull/31104#discussion_r775749309.
So now I believe that this is case is handled by your code of cache for storage because getting key from cache and creating a value holder object is done under mutex. And getting/setting keys in cache always goes throw that mutex.

Reply:
In our use case, there is and outside mutex which could avoid this case, but we cannot gurantee for other use cases

Reply:
Ok I agree, then may be instead of `bool canRelease` and `void release` have just `bool tryRelease()`?

Reply:
I just thought use `markRelease` for mark the element could be used anymore and `unmarkRelease` for rollback `removeOverflow`, but this may still have side effect

Reply:
When would we need to rollback it?

Reply:
It seems more likely what we need is a resource pool with key index and lru evict policy. The apis may are as following
```
// if could not insert into the cache, return nullptr
// if result is not null,  increase its reference
// use lru evict policy inside to evict elements which reference is 0
std::shared_ptr<Value> acquire(const Key & key,  LoadFunc && load_func);
// decrease its reference
void release(const Key & key, std::shared_ptr<Value> v);
```

Reply:
> When would we need to rollback it? May be just make implementation when we would not need rollback?

for example, we have elements : a,b,c, and a is releasabel, other two not. Now we want to insert d, but we cannot find enough space for d even after evict a. Maybe we just want to keep the cache elements as before. a, b and c are still in the cache

Reply:
> I still prefer a new implementation but not LRUCache

Yes, let's try to make a new one. I also start to think it will be better.